File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int countGoodNodes(TreeNode* root, int maxVal, int& res){
        
        if(root == NULL) return 0;
        
        res += root->val >= maxVal ? 1 : 0;
        
        countGoodNodes(root->left, max(root->val, maxVal), res);
        countGoodNodes(root->right, max(root->val, maxVal), res);
        
        return res;
    }
    
    int goodNodes(TreeNode* root) {
        
        int res = 0;
        return countGoodNodes(root, root->val, res);
    }
};

File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int helper(TreeNode root, int maxVal, int[] res){
        
        if(root == null) return 0;
        
        res[0] += root.val >= maxVal ? 1 : 0;
        
        helper(root.left, Math.max(maxVal, root.val), res);
        helper(root.right, Math.max(maxVal, root.val), res);
        
        return res[0];
    }
    
    public int goodNodes(TreeNode root) {
        int[] res = {0};
        return helper(root, root.val, res);
    }
}